### ISSUE TITLE
Batches freeze after committing +- 10 batches in a for loop

### ISSUE THREAD

**Issue by janpansa** [kreait/firebase-php#773]
### Describe the bug

I am trying to perform batch operations on Firestore.

I am splitting our batches into 490 items per batch. Firestore can only do 500 items per batch at a time according to their documentation.

Below is a snippet of my code:

```
   $batches = array_chunk($items, 490);
        $batchArr = [];
        foreach ($batches as $idx => $data) {
            $batch = $db->batch();
            foreach ($data as $datum) {
                $document = $db->collection('accounts')
                    ->document($accountId)
                    ->collection('items')
                    ->document($datum['itemId']);
               
                $batch->set($document, $datum, ['merge' => true]);
            }
            array_push($batchArr, $batch);
        }
        foreach ($batchArr as $batch) {
            if (!$batch->isEmpty()) {
                $batch->commit();
            } else {
                Log::debug('skipping empty batch');
            }
        }
        
```

The above code works 100% for smaller batches, e.g. 3 or 5 batches with 490 set operations within each one.

However, as soon as you have more than 10 batches, the loop just freezes without error and eventually times out.

I think what is happening is that it is not waiting for the batches to finish committing and just continuing the for loop, thus breaking the firestore rule of more than 500 items being committed per second.

According to Google's own documentation for PHP, batches have been deprecated and they recommend using bulk Writer.

See here -> https://googleapis.github.io/google-cloud-php/#/docs/cloud-firestore/v1.27.0/firestore/writebatch

and here is the bulkWriter class -> https://googleapis.github.io/google-cloud-php/#/docs/cloud-firestore/v1.27.0/firestore/bulkwriter

So we have also tried that, however, when using the bulkWriter, it seems the method does not exist within the kreait php sdk for firebase, because I get this error:

```
[2023-03-05 20:11:47] staging.ERROR: Call to undefined method Kreait\Firebase\Firestore::bulkWriter() {"exception":"[object] (Error(code: 0): Call to undefined method Kreait\\Firebase\\Firestore::bulkWriter() at /var/www/app/Services/SyncService/SyncService.php:706)
```

I have also noticed that this (writing/comitting multiple batches in sequence freezes) has been a reported issue within the nodejs sdk for firebase and was fixed in 2019, but not sure if it is related to this one.

### Installed packages

```json
{
  "name": "accounts/accountProgram",
  "type": "project",
  "description": "account program backend",
  "keywords": [],
  "license": "MIT",
  "require": {
    "php": "^8.1.0",
    "bpuig/laravel-subby": "^7.0",
    "calcinai/xero-php": "^2.5",
    "dealerinventory/client": "^1.22",
    "fruitcake/laravel-cors": "^2.0.5",
    "google/cloud-firestore": "^1.24",
    "guzzlehttp/guzzle": "^7.2",
    "kreait/laravel-firebase": "^4.1",
    "laravel/framework": "^9.0",
    "laravel/sanctum": "^2.14",
    "laravel/tinker": "^2.7",
    "mtownsend/progress": "^2.0",
    "predis/predis": "^1.1",
    "reliese/laravel": "^1.1",
    "rinvex/countries": "^8.1",
    "sentry/sentry-laravel": "^2.13",
    "spatie/laravel-fractal": "^6.0",
    "spatie/laravel-health": "^1.8",
    "spatie/laravel-permission": "^5.5",
    "spatie/laravel-sluggable": "^3.0",
    "vinkla/hashids": "^10.0"
  },
  "require-dev": {
    "roave/security-advisories": "dev-latest",
    "barryvdh/laravel-ide-helper": "^2.12",
    "brianium/paratest": "^6.4",
    "codedungeon/phpunit-result-printer": "^0.30",
    "fakerphp/faker": "^1.9.1",
    "friendsofphp/php-cs-fixer": "^3.6",
    "kitloong/laravel-migrations-generator": "^5.2",
    "mockery/mockery": "^1.4.4",
    "nunomaduro/collision": "^6.1",
    "nunomaduro/larastan": "^2.0",
    "osteel/openapi-httpfoundation-testing": "^0.7",
    "phpmd/phpmd": "dev-master",
    "phpstan/phpstan": "^1.4",
    "phpunit/phpunit": "^9.5.10",
    "spatie/laravel-ray": "^1.29",
    "squizlabs/php_codesniffer": "^3.6",
    "thedoctor0/laravel-factory-generator": "^1.2"
  },
  "autoload": {
    "files": [
      "helpers.php"
    ],
    "psr-4": {
      "App\\": "app/",
      "Database\\Factories\\": "database/factories/",
      "Database\\Seeders\\": "database/seeders/"
    }
  },
  "autoload-dev": {
    "psr-4": {
      "Tests\\": "tests/"
    }
  },
  "scripts": {
    "refresh": [
      "@php artisan config:clear",
      "@php artisan migrate:fresh --seed"
    ],
    "php-cs-fixer": "php-cs-fixer --config=.php-cs-fixer.php fix",
    "phpcs": "phpcs --standard=PSR1,PSR2 --exclude=Generic.Files.LineLength app/",
    "phpstan": "phpstan analyse --memory-limit 512M",
    "post-autoload-dump": [
      "Illuminate\\Foundation\\ComposerScripts::postAutoloadDump",
      "@php artisan package:discover --ansi"
    ],
    "post-update-cmd": [
      "@php artisan vendor:publish --tag=laravel-assets --ansi --force"
    ],
    "post-root-package-install": [
      "@php -r \"file_exists('.env') || copy('.env.example', '.env');\""
    ],
    "post-create-project-cmd": [
      "@php artisan key:generate --ansi"
    ]
  },
  "extra": {
    "laravel": {
      "dont-discover": []
    }
  },
  "config": {
    "optimize-autoloader": true,
    "preferred-install": "dist",
    "sort-packages": true
  },
  "minimum-stability": "dev",
  "prefer-stable": true
}
```

### PHP version and extensions

```shell
composer               2.3.10   Composer package
composer-plugin-api    2.3.0    The Composer Plugin API
composer-runtime-api   2.2.2    The Composer Runtime API
ext-bcmath             8.1.8    The bcmath PHP extension
ext-bz2                8.1.8    The bz2 PHP extension
ext-calendar           8.1.8    The calendar PHP extension
ext-ctype              8.1.8    The ctype PHP extension
ext-curl               8.1.8    The curl PHP extension
ext-date               8.1.8    The date PHP extension
ext-dom                20031129 The dom PHP extension
ext-exif               8.1.8    The exif PHP extension
ext-ffi                8.1.8    The FFI PHP extension
ext-fileinfo           8.1.8    The fileinfo PHP extension
ext-filter             8.1.8    The filter PHP extension
ext-ftp                8.1.8    The ftp PHP extension
ext-gd                 8.1.8    The gd PHP extension
ext-gettext            8.1.8    The gettext PHP extension
ext-gmp                8.1.8    The gmp PHP extension
ext-grpc               1.45.0   The grpc PHP extension
ext-hash               8.1.8    The hash PHP extension
ext-iconv              8.1.8    The iconv PHP extension
ext-igbinary           3.2.6    The igbinary PHP extension
ext-intl               8.1.8    The intl PHP extension
ext-json               8.1.8    The json PHP extension
ext-libxml             8.1.8    The libxml PHP extension
ext-mbstring           8.1.8    The mbstring PHP extension
ext-mysqli             8.1.8    The mysqli PHP extension
ext-mysqlnd            0        The mysqlnd PHP extension (actual version: mysqlnd 8.1.8)
ext-openssl            8.1.8    The openssl PHP extension
ext-pcntl              8.1.8    The pcntl PHP extension
ext-pcre               8.1.8    The pcre PHP extension
ext-pdo                8.1.8    The PDO PHP extension
ext-pdo_mysql          8.1.8    The pdo_mysql PHP extension
ext-pdo_sqlite         8.1.8    The pdo_sqlite PHP extension
ext-phar               8.1.8    The Phar PHP extension
ext-posix              8.1.8    The posix PHP extension
ext-protobuf           3.21.1   The protobuf PHP extension
ext-readline           8.1.8    The readline PHP extension
ext-redis              5.3.7    The redis PHP extension
ext-reflection         8.1.8    The Reflection PHP extension
ext-session            8.1.8    The session PHP extension
ext-shmop              8.1.8    The shmop PHP extension
ext-simplexml          8.1.8    The SimpleXML PHP extension
ext-sockets            8.1.8    The sockets PHP extension
ext-sodium             8.1.8    The sodium PHP extension
ext-spl                8.1.8    The SPL PHP extension
ext-sqlite3            8.1.8    The sqlite3 PHP extension
ext-sysvmsg            8.1.8    The sysvmsg PHP extension
ext-sysvsem            8.1.8    The sysvsem PHP extension
ext-sysvshm            8.1.8    The sysvshm PHP extension
ext-tokenizer          8.1.8    The tokenizer PHP extension
ext-xml                8.1.8    The xml PHP extension
ext-xmlreader          8.1.8    The xmlreader PHP extension
ext-xmlwriter          8.1.8    The xmlwriter PHP extension
ext-xsl                8.1.8    The xsl PHP extension
ext-zend-opcache       8.1.8    The Zend OPcache PHP extension
ext-zip                1.19.5   The zip PHP extension
ext-zlib               8.1.8    The zlib PHP extension
lib-bz2                1.0.8    The bz2 library
lib-curl               7.74.0   The curl library
lib-curl-libssh2       1.9.0    curl libssh2 version
lib-curl-openssl       1.1.1.14 curl OpenSSL version (1.1.1.14)
lib-curl-zlib          1.2.11   curl zlib version
lib-date-timelib       2021.12  date timelib version
lib-date-zoneinfo      0        zoneinfo ("Olson") database for date
lib-fileinfo-libmagic  540      fileinfo libmagic version
lib-gd                 2.3.3    The gd library
lib-gmp                6.2.1    The gmp library
lib-iconv              2.31     The iconv library
lib-icu                67.1     The ICU unicode and globalization support library
lib-icu-cldr           37       ICU CLDR project version
lib-icu-unicode        13.0.0   ICU unicode version
lib-libsodium          1.0.18   The libsodium library
lib-libxml             2.9.14   libxml library version
lib-libxslt            1.1.34   The libxslt library
lib-libxslt-libxml     2.9.10   libxml version libxslt is compiled against
lib-mbstring-libmbfl   1.3.2    mbstring libmbfl version
lib-mbstring-oniguruma 6.9.6    mbstring oniguruma version
lib-openssl            1.1.1.14 OpenSSL 1.1.1n  15 Mar 2022
lib-pcre               10.40    The pcre library
lib-pcre-unicode       14.0.0   PCRE Unicode version support
lib-pdo_sqlite-sqlite  3.34.1   The pdo_sqlite-sqlite library
lib-sqlite3-sqlite     3.34.1   The sqlite3-sqlite library
lib-zip-libzip         1.7.3    The zip-libzip library
lib-zlib               1.2.11   The zlib library
php                    8.1.8    The PHP interpreter
php-64bit              8.1.8    The PHP interpreter, 64bit
php-ipv6               8.1.8    The PHP interpreter, with IPv6 support
```

### Steps to reproduce the issue.

Step 1 -> Use the batch method to create batches that can be comitted to firestore, e.g.:

```
   $batches = array_chunk($inventory, 490);
        $batchArr = [];
        foreach ($batches as $idx => $data) {
            $batch = $db->batch();
            foreach ($data as $datum) {
                $document = $db->collection('accounts')
                    ->document($accountId)
                    ->collection('items')
                    ->document($datum['itemId']);
               
                $batch->set($document, $datum, ['merge' => true]);
            }
            array_push($batchArr, $batch);
        }
        foreach ($batchArr as $batch) {
            if (!$batch->isEmpty()) {
                $batch->commit();
            } else {
                Log::debug('skipping empty batch');
            }
        }
        
```

Step 2 -> Add 490 items per batch and commit them in a for loop, as shown above.
Step 3 -> Run into error after about 10 - 15 batches.

### Error message/Stack trace

```shell
It just freezes without error
```

**Comment by jeromegamez** [comment]
The SDK provides only a [simple bridge to the official Firestore client](https://github.com/kreait/firebase-php/blob/7.x/src/Firebase/Firestore.php), so there are really no methods present in the Firestore class other than the bridge method.

I can't see from the code how you got to the `$db` variable or how and where you invoked the Firestore class provided by the SDK. Also, I'm surprised that it works for the first batches but not for the subsequent ones.

All I can say is that in order to get the Firestore Client from the official SDK, you can do

```php
$firestore = $factory->createFirestore();

$firestoreClient = $firestore->database();
```

Then, `$firestore->database()` will return an instance of `Google\Cloud\Firestore\FirestoreClient` that will then give you access to everything the official client provides.

tl;dr; Don't call methods on the SDKs' `Firestore` instance directly, but use the `FirestoreClient` from the official Firestore library.

I hope this helps  🤞🏻

**Comment by janpansa** [comment]
Thanks @jeromegamez for your response.

Sure thing, yes we are using it as you described.

I tested again just now. When trying to "set" 8000+ items, in batches of 450 items per batch, it just freezes after about 15 batches of commits. It works fine for smaller batches, e.g. 3000 items, but as soon as the items exceed like +- 8000 then it starts freezing.

I have even tried to wait 1 sec after each commit but same result.

Here is the same issue that was happening in the js firebase sdk -> https://github.com/firebase/firebase-js-sdk/issues/1437

**Comment by jeromegamez** [comment]
I understand the issue - unfortunately, I don't own the code related to Firestore so I don't think I can help resolve the issue from this (unofficial) SDK's perspective.

The issue you linked suggests that, at that time, it was a Firebase backend issue; perhaps it's the same in this case 🤔. In any case, I think it would be best if you raised the issue at https://github.com/googleapis/google-cloud-php - the engineers of the official libraries do have access to the Firebase/Google backend and can certainly tell you if this is a problem within the code of the Firestore Client or a service backend issue.

**Comment by janpansa** [comment]
@jeromegamez ah! That makes sense!

Thanks for your help, I will take it up with them.

Thanks!

**Comment by janpansa** [comment]
To anyone who might have the same issue in the future, I have raised the issue at "https://github.com/googleapis/google-cloud-php " here -> https://github.com/googleapis/google-cloud-php/issues/5927

**Comment by jeromegamez** [comment]
Thank you for sharing the link, mich appreciated! 🙏🏻
